lib/plugins/aws/deploy/index.js

```diff
@@ -1,62 +1,111 @@
-      generateArtifactDirectoryName,
-      mergeCustomProviderResources,
       setBucketName,
       cleanupS3Bucket,
       uploadArtifacts,
       updateStack,
-      monitorStack,
-      configureStack,
-      mergeIamTemplates
+      monitorStack
     );
 
+    // Define the internal lifecycle model
+    this.commands = {
+      aws: {
+        type: 'entrypoint',
+        commands: {
+          deploy: {
+            commands: {
+              deploy: {
+                lifecycleEvents: [
+                  'createStack',
+                  'uploadArtifacts',
+                  'updateStack',
+                ],
+              },
+              finalize: {
+                lifecycleEvents: [
+                  'cleanup',
+                ],
+              },
+            },
+          },
+        },
+      },
+    };
+
     this.hooks = {
-      'before:deploy:initialize': () => BbPromise.bind(this)
-        .then(this.validate),
 
-      'deploy:initialize': () => BbPromise.bind(this)
-        .then(this.configureStack),
+      'before:deploy:deploy': () => BbPromise.bind(this)
+        .then(() => this.serverless.pluginManager.spawn('aws:common:validate'))
+        .then(() => {
+          if (!this.options.package && !this.serverless.service.package.path) {
+            return BbPromise.bind(this)
+            .then(() => this.serverless.pluginManager.spawn('package'))
+            .then(this.extendedValidate);
+          }
+          return BbPromise.bind(this)
+          .then(() => this.serverless.pluginManager.spawn('aws:common:moveArtifactsToTemp'))
+          .then(this.extendedValidate);
+        }),
 
-      'deploy:setupProviderConfiguration': () => BbPromise.bind(this)
-        .then(this.createStack)
-        .then(this.mergeIamTemplates),
+      // Deploy outer lifecycle
+      'deploy:deploy': () => BbPromise.bind(this)
+        .then(() => this.serverless.pluginManager.spawn('aws:deploy:deploy')),
 
-      'before:deploy:compileFunctions': () => BbPromise.bind(this)
-        .then(this.generateArtifactDirectoryName),
+      'deploy:finalize': () => this.serverless.pluginManager.spawn('aws:deploy:finalize'),
 
-      'deploy:deploy': () => BbPromise.bind(this)
-        .then(this.mergeCustomProviderResources)
+      // Deploy deploy inner lifecycle
+      'aws:deploy:deploy:createStack': () => BbPromise.bind(this)
+        .then(this.createStack),
+
+      'aws:deploy:deploy:uploadArtifacts': () => BbPromise.bind(this)
         .then(this.setBucketName)
-        .then(this.uploadArtifacts)
-        .then(this.updateStack)
+        .then(this.uploadArtifacts),
+
+      'aws:deploy:deploy:updateStack': () => BbPromise.bind(this)
+        .then(this.updateStack),
+
+      // Deploy finalize inner lifecycle
+      'aws:deploy:finalize:cleanup': () => BbPromise.bind(this)
         .then(this.cleanupS3Bucket)
         .then(() => {
-          if (this.options.noDeploy) this.serverless.cli.log('Did not deploy due to --noDeploy');
```

Comment:
Isn't --noDeploy already deprecated and removed since 1.9 or 1.10?

As this is a special case I would even suggest to tell users to switch to the package command instead of --noDeploy when switching to the new version. In general noDeploy does not make any sense anymore.

Another approach would be to add the noDeploy check to package:finalize which is called anyway